lib/cache/role.go

```diff
@@ -0,0 +1,156 @@
+
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/api/client/proto"
+	"github.com/gravitational/teleport/api/defaults"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/services"
+)
+
+func newRoleCollection(a services.Access, w types.WatchKind) (*collection[types.Role], error) {
+	if a == nil {
+		return nil, trace.BadParameter("missing parameter Access")
+	}
+
+	return &collection[types.Role]{
+		store: newStore(map[string]func(types.Role) string{
+			"name": func(r types.Role) string {
+				return r.GetName()
+			},
+		}),
+		fetcher: func(ctx context.Context, loadSecrets bool) ([]types.Role, error) {
+			return a.GetRoles(ctx)
+		},
+		headerTransform: func(hdr *types.ResourceHeader) types.Role {
+			return &types.RoleV6{
+				Kind:    types.KindRole,
+				Version: types.V7,
+				Metadata: types.Metadata{
+					Name: hdr.Metadata.Name,
+				},
+			}
+		},
+		watch: w,
+	}, nil
+}
+
+// GetRoles is a part of auth.Cache implementation
+func (c *Cache) GetRoles(ctx context.Context) ([]types.Role, error) {
+	ctx, span := c.Tracer.Start(ctx, "cache/GetRoles")
+	defer span.End()
+
+	rg, err := acquireReadGuard(c, c.collections.roles)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	defer rg.Release()
+
+	if !rg.ReadCache() {
+		roles, err := c.Config.Access.GetRoles(ctx)
+		return roles, trace.Wrap(err)
+	}
+
+	roles := make([]types.Role, 0, rg.store.len())
+	for r := range rg.store.resources("name", "", "") {
+		roles = append(roles, r.Clone())
+	}
+
+	return roles, nil
+}
+
+// ListRoles is a paginated role getter.
+func (c *Cache) ListRoles(ctx context.Context, req *proto.ListRolesRequest) (*proto.ListRolesResponse, error) {
+	ctx, span := c.Tracer.Start(ctx, "cache/ListRoles")
+	defer span.End()
+
+	rg, err := acquireReadGuard(c, c.collections.roles)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	defer rg.Release()
+
+	if !rg.ReadCache() {
+		resp, err := c.Config.Access.ListRoles(ctx, req)
+		return resp, trace.Wrap(err)
+	}
+
+	// Adjust page size, so it can't be too large.
+	pageSize := int(req.Limit)
+	if pageSize <= 0 || pageSize > defaults.DefaultChunkSize {
+		pageSize = defaults.DefaultChunkSize
+	}
+
+	var resp proto.ListRolesResponse
+	for r := range rg.store.resources("name", req.StartKey, "") {
+		rv6, ok := r.(*types.RoleV6)
+		if !ok {
+			continue
+		}
+
+		if req.Filter != nil && !req.Filter.Match(rv6) {
+			continue
+		}
+
+		if len(resp.Roles) == pageSize {
+			resp.NextKey = r.GetName()
+			break
+		}
+
+		resp.Roles = append(resp.Roles, r.Clone().(*types.RoleV6))
+
+	}
+	return &resp, nil
+}
+
+// GetRole is a part of auth.Cache implementation
+func (c *Cache) GetRole(ctx context.Context, name string) (types.Role, error) {
+	ctx, span := c.Tracer.Start(ctx, "cache/GetRole")
+	defer span.End()
+
+	rg, err := acquireReadGuard(c, c.collections.roles)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	defer rg.Release()
+
+	if !rg.ReadCache() {
```

Comment:
I looked for the documentation of `ReadCache`, and I found no hint of what this function does. The field that it returns is also undocumented. Can you please add a comment there? After looking at the code, it looks like `ReadCache` is supposed to tell me whether to read the cache at all or just skip to the underlying service, but it's not really obvious.

TBH, the `readGuard` concept is also a bit hard to understand: this struct also has no comments, and the `acquireReadGuard` function documentation only says that the guard "may be used to determine how a read cache should operate", which doesn't really help in understanding what "how should operate" exactly means.

Reply:
Thanks!